File: ExpectedPerformance/TurnOnCurve.cxx
// This code plots the turn on curve where the entries are weighted
// following the lumi inst profile in a single fill.
// The output is the variance of the derivative distribution

#include"tdrstyle.C"
#include<cmath>

void TurnOnCurve() {

    gStyle->SetOptStat(0);
  
    setTDRStyle();

    std::ifstream in_file;
    std::vector<float> transparency;
    std::vector<float> lumi_inst;
    std::vector<float> lumi_int;
    std::vector<float> time; //timestamp vector
    float x,y,z, t;

    // Load real transparency data, with lumi inst and lumi in fill
//     in_file.open("iring24.txt");
    in_file.open("iring25.txt");
//     in_file.open("iring26.txt");
    if (!in_file) {
        std::cout << "error" << std::endl;
        exit(1);
    }

    while (in_file >> x >> y >> z >> t) {
        transparency.push_back(x);
        lumi_inst.push_back(y);
        lumi_int.push_back(z);
        time.push_back(t);
    }

    int transparency_size = transparency.size();

    int nbin = 80;
    float minimo = 29;
    float massimo = 31;
    float threshold = 30;
    float delta_value = (massimo-minimo)/nbin;

    TCanvas* cc_turn_on = new TCanvas ("cc_turn_on", "", 800, 600);
    TH1F *h_correction = new TH1F("h_correction", "", nbin, minimo, massimo);
    TH1F *h_real = new TH1F ("h_real", "", nbin, minimo, massimo);
    TH1F *h_ideal = new TH1F ("h_ideal", "", nbin, minimo, massimo);

    // The function has been divided into two parts just for simplicity
    TF1* lumi_int_function = new TF1 ("lumi_int_function", "[0]*exp(-[1]*x)+(1-[0])*exp([2]*x)");
    TF1* lumi_inst_function = new TF1 ("lumi_inst_function", "[0]*exp(-[1]*(x-[3]))+(1-[0])*exp([2]*(x-[3]))");

    // Fit parameters obtained from fit @ iring25    
    lumi_int_function->SetParameter(0, 9.93e-1);
    lumi_int_function->SetParameter(1, 3.87e-2);
    lumi_int_function->SetParameter(2, 3.22);

    lumi_inst_function->SetParameter(0, 7.89);
    lumi_inst_function->SetParameter(1, 2.71);
    lumi_inst_function->SetParameter(2, -3.02);

    gRandom->SetSeed();

//     float lumi_min = 0.000100052299505594;
//     float lumi_max = 0.000390878557887418;
//     float delta_lumi = lumi_max - lumi_min;
    float lumi_inst_0;
    float lumi_int_0;
   
    
    // Fill histograms to get turn on curves
    for (int ibin = 0; ibin < nbin+1; ibin++) {
	float sum_weight = 0.;
        float value = minimo + (ibin+0.5)*delta_value;
//         std::cout << ibin << "/" << nbin << std::endl;
        for (int i = 0; i < transparency_size ; i++) {
            if (transparency[i] == 1.) {
                lumi_inst_0 = lumi_inst[i];
                lumi_inst_function->SetParameter(3, lumi_inst_0);
                lumi_int_0 = lumi_int[i];
            }
            y = lumi_inst[i] ;
            z = lumi_int[i] - lumi_int_0; // AM : FIX  
            float correction = (lumi_int_function->Eval(z))*(lumi_inst_function->Eval(y)); // compute correction 
            float value_smeared = value*transparency[i]/correction; // value w/ correction
            float value_smeared_real = value*transparency[i]; // value w/out correction
            sum_weight += lumi_inst_0/y;
            if (value_smeared > threshold) {
                h_correction->Fill(value, lumi_inst_0/y);
            }
            if (value_smeared_real > threshold) {
                h_real->Fill(value, lumi_inst_0/y);
            }
	    if (value > threshold) {
		h_ideal->Fill(value, lumi_inst_0/y);
	    }
        }

	double scale_real = h_real->GetBinContent(ibin)/sum_weight;
	double scale_correction = h_correction->GetBinContent(ibin)/sum_weight;
	double scale_ideal = h_ideal->GetBinContent(ibin)/sum_weight;

	h_real->SetBinContent(ibin, scale_real);
	h_correction->SetBinContent(ibin, scale_correction);
	h_ideal->SetBinContent(ibin, scale_ideal);
    }

    h_ideal->SetLineWidth(1.);
    h_ideal->SetLineColor(kBlack);
    h_ideal->SetStats(0);

    h_ideal->Draw("histo");
    h_ideal->GetXaxis()->SetTitle("Energy [GeV]");
    h_ideal->GetYaxis()->SetTitle("Efficiency");

    h_real->SetLineWidth(2.);
    h_real->SetLineColor(kBlue);
    h_real->SetStats(0);

    h_real->Draw("histo same");
    h_real->GetXaxis()->SetTitle("Energy [GeV]");
    h_real->GetYaxis()->SetTitle("Efficiency");
  
    h_correction->SetLineWidth(2.);
    h_correction->SetLineColor(kRed);
    h_correction->SetStats(0);

    h_correction->Draw("histo same");
    h_correction->GetXaxis()->SetTitle("Energy [GeV]");
    h_correction->GetYaxis()->SetTitle("Efficiency");

    TLegend *legend = new TLegend();
    legend->AddEntry(h_real,"Without correction");
    legend->AddEntry(h_correction,"With correction");
    legend->AddEntry(h_ideal, "Ideal");
    legend->Draw();

    //
    // Derivatives of the turn on curves
    //
    Double_t x_real[nbin], y_real[nbin], x_correction[nbin], y_correction[nbin], d_real[nbin], d_correction[nbin];

    for (int i = 1; i < nbin+1; i++) {
        x_real[i] = h_real->GetBinCenter(i);
        y_real[i] = h_real->GetBinContent(i);

        x_correction[i] = h_correction->GetBinCenter(i);
        y_correction[i] = h_correction->GetBinContent(i);
    }

    for (int i = 1; i < nbin+1; i++) {
        d_real[i] = (y_real[i]-y_real[i-1])/(x_real[i]-x_real[i-1]);
        d_correction[i] = (y_correction[i]-y_correction[i-1])/(x_correction[i]-x_correction[i-1]);
    }

    Double_t point_real[nbin-8], point_correction[nbin-8];
    double sum_real = 0.;
    double sum_correction = 0.;

    for (int i = 2; i < nbin; i++) {
        sum_real = 0.;
        sum_correction = 0.;
        sum_real = d_real[i-1]+d_real[i]+d_real[i+1]; //3 points average to smooth a little
        sum_correction = d_correction[i-1]+d_correction[i]+d_correction[i+1];
        point_real[i] = sum_real/3;
        point_correction[i] = sum_correction/3;
    }


    TCanvas *c_point = new TCanvas("c_point", "", 800, 700);
    TGraph *real_point = new TGraph(nbin, x_real, point_real);
    TGraph *correction_point = new TGraph(nbin, x_correction, point_correction);
    TMultiGraph *mg2 = new TMultiGraph();

    real_point->SetMarkerSize(1.);
    real_point->SetMarkerStyle(20);
    real_point->SetMarkerColor(kBlue);
    real_point->SetLineColor(kBlue);
    real_point->SetLineWidth(2);
    mg2->Add(real_point);

    correction_point->SetMarkerSize(1.);
    correction_point->SetMarkerStyle(20);
    correction_point->SetMarkerColor(kRed);
    correction_point->SetLineColor(kRed);
    correction_point->SetLineWidth(2);
    mg2->Add(correction_point);

    mg2->Draw("AP");
    mg2->GetXaxis()->SetLimits(29,31);
    mg2->SetMinimum(-1e-3);
    mg2->GetXaxis()->SetTitle("Energy [GeV]");
    mg2->GetXaxis()->SetLabelSize(.045);
    mg2->GetYaxis()->SetTitle("Efficiency derivative [GeV^{-1}]");

    TLegend *p_legend = new TLegend();
    p_legend->AddEntry(real_point,"Without correction");
    p_legend->AddEntry(correction_point,"With correction");
    p_legend->Draw();

    sum_real = 0.;
    sum_correction = 0.;

    for (int i = 1; i < nbin+1; i++) {
        sum_real += d_real[i];
        sum_correction += d_correction[i];
    }
    double mean_real = sum_real/nbin;
    double mean_correction = sum_correction/nbin;

    // Variance computation
    double var_real = 0.;
    double var_correction = 0.;
    double weight_real = 0.;
    double weight_correction = 0.;
    for (int i = 1; i < nbin+1; i++) {
        var_real += point_real[i]*(x_real[i]-30.)*(x_real[i]-30.);
        weight_real += point_real[i];

        var_correction += point_correction[i]*(x_correction[i]-30.)*(x_correction[i]-30.);
        weight_correction += point_correction[i];
    }
    var_real = var_real/weight_real;
    var_correction = var_correction/weight_correction;
    std::cout << "Without correction: " << var_real << std::endl;
    std::cout << "With correction: " << var_correction << std::endl;

    
    //
    // Draw transparency and its prediction
    //
    
    TCanvas* cc_transp = new TCanvas ("cc_transp", "", 800, 600);
    
    Double_t x_time[transparency_size], y_transp[transparency_size], y_transp_predict[transparency_size];
    
    for (int i = 0; i < transparency_size ; i++) {
      if (transparency[i] == 1.) {
        lumi_inst_0 = lumi_inst[i];
        lumi_inst_function->SetParameter(3, lumi_inst_0);
        lumi_int_0 = lumi_int[i];
//         std::cout << "  lumi_inst[" << i << "]  = " <<  lumi_inst[i]  <<  " lumi_int[" << i << "] = " << lumi_int[i] << std::endl;
      }
      y = lumi_inst[i] ;
      z = lumi_int[i] - lumi_int_0; // AM : FIX  
      float correction = (lumi_int_function->Eval(z))*(lumi_inst_function->Eval(y)); // compute correction 
      
      y_transp_predict[i] = correction;
      y_transp[i] = transparency[i];
      x_time[i] = time[i];
    }
    
    TGraph* gr_tr = new TGraph(transparency_size,x_time,y_transp);
    gr_tr->SetMarkerSize(1.);
    gr_tr->SetMarkerStyle(20);
    gr_tr->SetMarkerColor(kRed);
    gr_tr->SetLineColor(kRed);
    gr_tr->Draw("APL");
    gr_tr->GetXaxis()->SetTimeDisplay(0);
    
    TGraph* gr_tr_predict = new TGraph(transparency_size,x_time,y_transp_predict);
    gr_tr_predict->SetMarkerSize(1.);
    gr_tr_predict->SetMarkerStyle(21);
    gr_tr_predict->SetMarkerColor(kBlue);
    gr_tr_predict->SetLineColor(kBlue);
    gr_tr_predict->Draw("PL");
    gr_tr_predict->GetXaxis()->SetTimeDisplay(0);
    
    cc_transp->SetGrid();
    
    
}

File: ExpectedPerformance/TurnOnCurve.py
#!/usr/bin/env python
# coding: utf-8

# In[ ]:


import ROOT
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from scipy.optimize import curve_fit
from scipy import special


# In[ ]:


def transp_func(data,a,b,c,d,e,f):
    x = data[0]
    y = data[1]
    y_0 = data[2]
    return (a*np.exp(-b*x)+(1-a)*np.exp(c*x))*(d*np.exp(-e*(y-y_0))+(1-d)*np.exp(f*(y-y_0)))


# In[ ]:


def transp_func_2(data,a,b,c,d,e,f):
    x = data[0]
    y = data[1]
    y_0 = data[2]
    return (a*np.exp(-b*x)+(1-a)*np.exp(c*x))*(d*np.exp(-e*y)+(1-d)*np.exp(f*y))/(d*np.exp(-e*y_0)+(1-d)*np.exp(f*y_0))


# In[ ]:


data_folder=("/home/alessandro/Scrivania/University/ML4ECAL/Transparency/DataPreparation")


# In[ ]:


#Read metadata
metadata = pd.read_csv(f"{data_folder}/fill_metadata_2017_10min.csv")


# In[ ]:


#Load data
data = np.load(f"{data_folder}/iring_25.npy")
data_test = np.load(f"{data_folder}/iring_26.npy")
data_test_2 = np.load(f"{data_folder}/iring_24.npy")

data_df = pd.DataFrame(data)
data_df_test = pd.DataFrame(data_test)
data_df_test_2 = pd.DataFrame(data_test_2)


# In[ ]:


#Mean transparency in iRing
mean = []
mean_test = []
mean_test_2 = []
for i in range(0, len(data_df.axes[1])):
    mean = np.append(mean, np.mean(data_df[i]))
    mean_test = np.append(mean_test, np.mean(data_df_test[i]))
    mean_test_2 = np.append(mean_test_2, np.mean(data_df_test_2[i]))


# In[ ]:


#Filter data and metadata
mean = mean[mean != -1]
mean_test = mean_test[mean_test != -1]
mean_test_2 = mean_test_2[mean_test_2 != -1]
metadata = metadata.iloc[:len(mean)][mean != -1]


# In[ ]:


fill = metadata["fill_num"].unique()
fill= fill[fill != 0]


# In[ ]:


metadata_fill = metadata[metadata.fill_num.isin(fill)]
metadata_fill = metadata_fill[(metadata_fill.lumi_inst >= 0.0001*1e9) & (metadata_fill.lumi_inst <= 0.0004*1e9) & (metadata_fill.lumi_in_fill >= 0.1*1e9)]


# In[ ]:


fill_num = metadata_fill.fill_num.unique()
transp_fill = []
lumi_inst_0 = []
lumi_int_0 = []

for k in fill_num:
    df = metadata_fill[metadata_fill.fill_num == k]
    transp = [mean[i] for i in df.index.values]
    transp = transp/transp[0]
    transp_fill = np.append(transp_fill, transp)
    a = np.empty(np.size(transp))
    b = np.empty(np.size(transp))
    a.fill(df['lumi_inst'].iloc[0])
    b.fill(df['lumi_int'].iloc[0])
    lumi_inst_0 = np.append(lumi_inst_0, a)
    lumi_int_0 = np.append(lumi_int_0, b)


# In[ ]:


par_1, pcov_1 = curve_fit(transp_func, [metadata_fill.lumi_in_fill*(1e-9), metadata_fill.lumi_inst*(1e-9), lumi_inst_0*(1e-9)], transp_fill, maxfev=5000)
par_2, pcov_2 = curve_fit(transp_func_2, [metadata_fill.lumi_in_fill*(1e-9), metadata_fill.lumi_inst*(1e-9), lumi_inst_0*(1e-9)], transp_fill, maxfev=5000)


# In[ ]:


nbin = 600
minimo = 0
massimo = 60
threshold = 30
delta_value = (massimo-minimo)/nbin


# In[ ]:


nEvents = 1000


# # Single Fill

# In[ ]:


fill = input()


# In[ ]:


selected_metadata = metadata[metadata.fill_num == int(fill)]
selected_transp = [mean[i] for i in selected_metadata.index.tolist()]


# In[ ]:


lumi_in_fill = selected_metadata.lumi_in_fill.to_numpy()
lumi_inst = selected_metadata.lumi_inst.to_numpy()
lumi_inst_0 = np.empty(np.size(selected_transp))
lumi_inst_0.fill(lumi_inst[0])


# In[ ]:


c = ROOT.TCanvas("cc_turn_on", "", 800, 700)
hist0 = ROOT.TH1F("real", "", nbin, minimo, massimo)


# In[ ]:


for ibin in range(0, nbin):
    value = minimo+(ibin+0.2)*delta_value
    for iEvent in range(0, nEvents):
        for i in range(0, np.size(selected_transp)):
            value_smeared = value*selected_transp[i]
            if value_smeared > threshold:
                hist0.Fill(value)


# In[ ]:


hist0.Scale(1./(nEvents*np.size(selected_transp)))
  
hist0.SetLineWidth(2)
hist0.SetLineColor(632)
  
hist0.Draw("histo")
hist0.GetXaxis().SetTitle("Energy [GeV]")
hist0.GetYaxis().SetTitle("Efficiency")


# In[ ]:


c.Draw()
vertical_line = ROOT.TLine(threshold, 0.0, threshold, 1.1)
vertical_line.Draw()
#c1.SaveAs("h_turn_on.png")


# ## $f_1$

# In[ ]:


c0 = ROOT.TCanvas("cc_turn_on", "", 800, 700)
hist = ROOT.TH1F("f_1", "", nbin, minimo, massimo)


# In[ ]:


transparency = transp_func([lumi_in_fill*(1e-9), lumi_inst*(1e-9), lumi_inst_0*(1e-9)], *par_1)


# In[ ]:


for ibin in range(0, nbin):
    value = minimo+(ibin+0.2)*delta_value
    for iEvent in range(0, nEvents):
        for i in range(0, np.size(selected_transp)):
            value_smeared = value*transparency[i]
            if value_smeared > threshold:
                hist.Fill(value)


# In[ ]:


hist.Scale(1./(nEvents*np.size(selected_transp)))
  
hist.SetLineWidth(2)
hist.SetLineColor(632)
  
hist.Draw("histo")
hist.GetXaxis().SetTitle("Energy [GeV]")
hist.GetYaxis().SetTitle("Efficiency")


# In[ ]:


c0.Draw()
vertical_line = ROOT.TLine(threshold, 0.0, threshold, 1.1)
vertical_line.Draw()
#c1.SaveAs("h_turn_on.png")


# ## $f_2$

# In[ ]:


c1 = ROOT.TCanvas("cc_turn_on_2", "", 800, 700)
hist2 = ROOT.TH1F("f_2", "", nbin, minimo, massimo)


# In[ ]:


transparency = transp_func_2([lumi_in_fill[i]*(1e-9), lumi_inst[i]*(1e-9), lumi_inst_0[i]*(1e-9)], *par_2)


# In[ ]:


for ibin in range(0, nbin):
    value = minimo+(ibin+0.2)*delta_value
    for iEvent in range(0, nEvents):
        for i in range(0, np.size(selected_transp)):
            value_smeared = value*transparency
            if value_smeared > threshold:
                hist2.Fill(value)


# In[ ]:


hist2.Scale(1./(nEvents*np.size(selected_transp)))
  
hist2.SetLineWidth(2)
hist2.SetLineColor(632)
  
hist2.Draw("histo")
hist2.GetXaxis().SetTitle("Energy [GeV]")
hist2.GetYaxis().SetTitle("Efficiency")


# In[ ]:


c1.Draw()
vertical_line = ROOT.TLine(threshold, 0.0, threshold, 1.1)
vertical_line.Draw()
#c1.SaveAs("h_turn_on.png")


# In[ ]:




